File: backend/Minimarket.Api/Controllers/SuppliersController.cs
using Microsoft.AspNetCore.Mvc;
using Minimarket.Api.DTOs;
using Minimarket.Api.Services;

namespace Minimarket.Api.Controllers;

[ApiController]
[Route("api/[controller]")]
public class SuppliersController(ISupplierService supplierService) : ControllerBase
{
    [HttpGet]
    public async Task<ActionResult<IReadOnlyCollection<SupplierDto>>> GetAll()
        => Ok(await supplierService.GetAllAsync());

    [HttpGet("{id:int}")]
    public async Task<ActionResult<SupplierDto>> GetById(int id)
    {
        var supplier = await supplierService.GetByIdAsync(id);
        return supplier is null ? NotFound() : Ok(supplier);
    }

    [HttpPost]
    public async Task<ActionResult<SupplierDto>> Create([FromBody] SaveSupplierDto dto)
    {
        var result = await supplierService.CreateAsync(dto);
        if (!result.Success)
        {
            return BadRequest(new { message = result.Error });
        }

        return CreatedAtAction(nameof(GetById), new { id = result.Supplier!.Id }, result.Supplier);
    }

    [HttpPut("{id:int}")]
    public async Task<ActionResult<SupplierDto>> Update(int id, [FromBody] SaveSupplierDto dto)
    {
        var result = await supplierService.UpdateAsync(id, dto);
        if (!result.Success)
        {
            return result.Error == "Proveedor no encontrado."
                ? NotFound(new { message = result.Error })
                : BadRequest(new { message = result.Error });
        }

        return Ok(result.Supplier);
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Delete(int id)
    {
        var result = await supplierService.DeleteAsync(id);
        if (!result.Success)
        {
            return NotFound(new { message = result.Error });
        }

        return NoContent();
    }
}

File: backend-java/minimarket-api-java/src/main/java/com/minimarket/api/controller/SuppliersController.java
package com.minimarket.api.controller;

import com.minimarket.api.dto.ApiMessageResponse;
import com.minimarket.api.dto.SaveSupplierDto;
import com.minimarket.api.dto.SupplierDto;
import com.minimarket.api.service.SupplierService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.List;

@RestController
@RequestMapping("/api/suppliers")
public class SuppliersController {

    private final SupplierService supplierService;

    public SuppliersController(SupplierService supplierService) {
        this.supplierService = supplierService;
    }

    @GetMapping
    public List<SupplierDto> getAll() {
        return supplierService.getAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<SupplierDto> getById(@PathVariable Integer id) {
        var supplier = supplierService.getById(id);
        return supplier == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(supplier);
    }

    @PostMapping
    public ResponseEntity<?> create(@RequestBody SaveSupplierDto dto) {
        var result = supplierService.create(dto);
        if (!result.success()) {
            return ResponseEntity.badRequest().body(new ApiMessageResponse(result.error()));
        }

        URI location = ServletUriComponentsBuilder.fromCurrentRequest()
            .path("/{id}")
            .buildAndExpand(result.data().id())
            .toUri();

        return ResponseEntity.created(location).body(result.data());
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(@PathVariable Integer id, @RequestBody SaveSupplierDto dto) {
        var result = supplierService.update(id, dto);
        if (!result.success()) {
            if ("Proveedor no encontrado.".equals(result.error())) {
                return ResponseEntity.status(404).body(new ApiMessageResponse(result.error()));
            }

            return ResponseEntity.badRequest().body(new ApiMessageResponse(result.error()));
        }

        return ResponseEntity.ok(result.data());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(@PathVariable Integer id) {
        var result = supplierService.delete(id);
        if (!result.success()) {
            return ResponseEntity.status(404).body(new ApiMessageResponse(result.error()));
        }

        return ResponseEntity.noContent().build();
    }
}
